pkg/kernel/networkservice/connectioncontextkernel/ipcontext/iprule/common.go

```diff
@@ -54,45 +55,72 @@ func create(ctx context.Context, conn *networkservice.Connection, tableIDs *Map,
 			return errors.WithStack(err)
 		}
 
-		for _, policy := range conn.Context.IpContext.Policies {
-			// Check if we already created required ip table
-			key := tableKey{
-				connId:   conn.GetId(),
-				from:     policy.From,
-				protocol: policy.Proto,
-				dstPort:  policy.DstPort,
-				srcPort:  policy.SrcPort,
-			}
-			tableID, ok := tableIDs.Load(key)
-			if !ok {
-				counter.Inc()
-				tableID = int(counter.Load())
+		ps, ok := tableIDs.Load(mechanism.GetNetNSURL())
+		if !ok {
+			ps = policies{
+				policies: make(map[int]*networkservice.PolicyRoute),
 			}
+			tableIDs.Store(mechanism.GetNetNSURL(), ps)
+		}
+		// Get policies to add and to remove
+		toAdd, toRemove := getPolicyDifferences(ps.policies, conn.Context.IpContext.Policies)
 
-			// If policy doesn't contain any route - add default
-			if len(policy.Routes) == 0 {
-				policy.Routes = append(policy.Routes, defaultRoute())
+		// Remove no longer existing policies
+		for tableID, policy := range toRemove {
+			if err := delRule(ctx, netlinkHandle, policy); err != nil {
+				return err
 			}
+			delete(ps.policies, tableID)
+			tableIDs.Store(mechanism.GetNetNSURL(), ps)
+		}
+		// Add new policies
+		for _, policy := range toAdd {
+			tableID := int(ps.counter.Inc())
 			for _, route := range policy.Routes {
 				if err := routeAdd(ctx, netlinkHandle, l, route, tableID); err != nil {
 					return err
 				}
 			}
-
-			if !ok {
-				// Check and delete old rules if they don't fit
-				_ = delOldRules(ctx, netlinkHandle, policy, tableID)
-				// Add new rule
-				if err := ruleAdd(ctx, netlinkHandle, policy, tableID); err != nil {
-					return err
-				}
-				tableIDs.Store(key, tableID)
+			if err := ruleAdd(ctx, netlinkHandle, policy, tableID); err != nil {
+				return err
 			}
+			ps.policies[tableID] = policy
+			tableIDs.Store(mechanism.GetNetNSURL(), ps)
```

Comment:
Can we add `ps` to `tableIDs` after loop?

Reply:
not sure, if for some reason, routeAdd/ruleAdd fails, the previously added rules/routes will be forgotten, so I think it is better to keep the tableIDs update whithin the loop.